Lint recursive application using default args, quick use of lint roller

som-snytt opened the pull request:
Revives retronym's warning for using default args when recursing, since you don't know if the default was intended or to thread the arg down the stack.

In this case paulp was the good angel on his shoulder.

https://github.com/retronym/scala/commit/3bfb6830

It would have caught [this one](https://github.com/scala/scala/commit/7d4109486b2266f8491d3473f43555dec6e996ee#diff-18ea56bad5c5d7a7b5af47b7131bebc1R1087), where the coder even says, "I shouldn't be using a boolean param this way."

I didn't devise a test for that code path; it's clearly wrong; but maybe fixing it breaks ~~someone's workaround~~ some tests.

This PR also partially addresses `jsig` in `Erasure` by adding a specialized version for the common recursive case, and otherwise adding args, including naming boolean literals, or `booterals` as I just typed.

Fixed up some stray unused imports.

I may add a commit for the infamous `higherKinds` import. ~~Or maybe just by always tagging it used? Where is the harm?~~

som-snytt committed:
Lint recursion with default args

som-snytt committed:
Syntax fixes

som-snytt committed:
Bad recursion with default args in interactive

The commit says: "While I agree that using a boolean argument
...isn't a great design...". The commit touched the recursive
call, so it's clear that this anti-pattern is highly pernicious.

som-snytt committed:
Avoid overload with defaults in specialize

som-snytt committed:
Avoid not specifying original is empty in Typer

som-snytt committed:
Mitigate recursion with default args in erasure

som-snytt committed:
Stray unused imports

som-snytt committed:
Reduce higherKinds import

som-snytt commented:
So close! One gold dot remaining!

som-snytt committed:
Lint HashMap

som-snytt committed:
Lint RedBlackTree

som-snytt committed:
Lint Vector

som-snytt committed:
Extend Java-defined exclusion for nullary override

StringBuilder.length overrides CharSequence.length().
Don't warn if there is a Java-defined method anywhere
among the overrides.

som-snytt committed:
Extend nullary override test

som-snytt committed:
Lint partest

som-snytt committed:
Lint repl-frontend

som-snytt committed:
Lint collection

som-snytt committed:
More lints

som-snytt committed:
Further lints

som-snytt commented:
I hope @retronym likes seeing his old efforts bear fruit. It starts with his lint rule, applies it over a few commits, then the rest is just junky linting that should be automated. Asking @hrhino because he will enjoy looking back at old paulpisms.

hrhino commented on src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:
```diff
@@ -1094,10 +1094,10 @@ trait Scanners extends ScannersCommon {
+    @inline private def isNumberSeparator(c: Char): Boolean = c == '_'
 
     @inline private def removeNumberSeparators(s: String): String =
-      if (s.indexOf('_') > 0) s.replaceAllLiterally("_", "") /*.replaceAll("'","")*/ else s
```
We could have had `1'000'000`? Now that `Symbol` is gone what's there left for poor `'`? Other than character literals, of course. I guess it's not so much gone as just quietly not reinvited.

som-snytt replied:
Ichoran wanted `1'000`. There's also `x'` for redefining `x`.

hrhino commented on src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala:
```diff
@@ -64,9 +64,9 @@ abstract class BCodeHelpers extends BCodeIdiomatic {
-    else name + nme.NAME_JOIN_STRING
+    val nameString = sym.javaSimpleName.toString
+    if (sym.isMixinConstructor) nameString
+    else nameString + nme.NAME_JOIN_STRING
```
Is it really `NAME_JOIN_STRING` if it's not joining names?

hrhino commented on src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:
```diff
@@ -1145,7 +1145,8 @@ object BTypes {
       scala.util.Sorting.quickSort(result)(Ordering.by(_._1))
-      result
+      import scala.collection.immutable.ArraySeq
+      ArraySeq.unsafeWrapArray(result)
```
Is the style local imports or FQNs? Pace that one "DOES NOT WORK in general" commit, `collection.immutable.ArraySeq` is a nice middle ground. Even more so if you `import collection._`, because who doesn't need a nice collection?

som-snytt replied:
FTR I prefer local import. On this scale of linting, I tried to avoid messing with imports in large scopes except for removing unused.

som-snytt replied:
I imported the unsafe method at top. It looks much cleaner.

hrhino commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1315,11 +1315,8 @@ abstract class RefChecks extends Transform {
-        ) + "\nClasses which cannot access %s %s %s.".format(
-          otherSym.decodedName, cannot, memberSym.decodedName)
+        s"""|${memberSym.fullLocationString}${comparison} references ${accessFlagsToString(otherSym)} ${otherSym}.
+            |Classes which cannot access ${otherSym.decodedName} ${cannot} ${memberSym.decodedName}.""".stripMargin
```
Cool. I keep forgetting that we have `format` over good old `s`/`f`.

som-snytt replied:
"I left my heart in good old `s`/`f`..." 🎶

hrhino commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -3714,7 +3714,11 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
-                  doTypedApply(tree, if (blockIsEmpty) fun else fun1, allArgs, mode, pt)
+                  def checkRecursive(res: Tree): Unit =
+                    if (settings.warnRecurseWithDefault && !res.isErroneous && context.owner.hasTransOwner(funSym))
+                      context.warning(res.pos, "Recursive call used default arguments.")
```
You could be a bit more helpful by doing something like `recursive call uses default arguments rather than the function's arguments`

som-snytt replied:
Right, because you think, "As opposed to what?" How about a, "Did you mean...?"  "Did you intend to thread the function's arguments down the stack?" Or only under `-Ysecond-guess`?

som-snytt replied:
Update the string from retronym minimalism.

hrhino commented on src/interactive/scala/tools/nsc/interactive/Picklers.scala:
```diff
@@ -42,9 +42,7 @@ trait Picklers { self: Global =>
-  private val sourceFilesSeen = new mutable.HashMap[AbstractFile, Array[Char]] {
-    override def default(key: AbstractFile) = Array()
-  }
+  private val sourceFilesSeen = mutable.HashMap.empty[AbstractFile, Array[Char]].withDefaultValue(Array.empty[Char])
```
I suppose there was a time before `withDefaultValue`?

hrhino commented on src/interactive/scala/tools/nsc/interactive/tests/core/AskCommand.scala:
```diff
@@ -121,7 +122,7 @@ trait AskTypeAt extends AskCommand {
   import compiler.Tree
 
-  protected def askLoadedTyped(source: SourceFile, keepLoaded: Boolean = false)(implicit reporter: Reporter): Response[Tree] = {
+  protected def askLoadedTyped(source: SourceFile, keepLoaded: Boolean = false)(implicit @unused reporter: Reporter): Response[Tree] = {
```
`implicit @unused` or `@unused implicit`? Why isn't this part of the style guide?

som-snytt replied:
💯

som-snytt replied:
It's part of the spec. `(implicit params...)` and `@annot id`. Less obvious with just one param.

hrhino commented on src/library/scala/collection/mutable/Buffer.scala:
```diff
@@ -227,5 +227,4 @@ object Buffer extends SeqFactory.Delegate[Buffer](ArrayBuffer)
 object IndexedBuffer extends SeqFactory.Delegate[IndexedBuffer](ArrayBuffer)
 
 /** Explicit instantiation of the `Buffer` trait to reduce class file size in subclasses. */
-@SerialVersionUID(3L)
```
I remember this lint. It was meant to victim-blame the victims of some SVUID bug I introduced, I think.

hrhino commented on src/reflect/scala/reflect/internal/Depth.scala:
```diff
@@ -31,6 +31,7 @@ final class Depth private (val depth: Int) extends AnyVal with Ordered[Depth] {
   override def toString = s"Depth($depth)"
 }
 
+@FunctionalInterface
```
Do we give effect to this?

som-snytt replied:
yes, for SAM consideration, because WWJD.

hrhino commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -98,9 +98,9 @@ trait Types
+  @unused private final val emptySymbolSet = immutable.Set.empty[Symbol]
 
-  private final val breakCycles = settings.breakCycles.value
+  @unused private final val breakCycles = settings.breakCycles.value
```
What are they for?

som-snytt replied:
I was just documenting.

hrhino commented on src/reflect/scala/reflect/internal/Variance.scala:
```diff
@@ -92,6 +92,7 @@ object Variance {
 
+  @FunctionalInterface
   trait Extractor[A]     { def apply(x: A): Variance }
   trait Extractor2[A, B] { def apply(x: A, y: B): Variance }
```
`Extractor2` resents being considered nonfunctional. I know there are people in the community for whom that is the greatest insult.

hrhino reviewed (approved):
I assume that that means, "look, but don't touch".

som-snytt commented:
I'll resolve the conflicts but would like @lrytz command decision because of the lint and the scope of files touched.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -458,9 +458,9 @@ abstract class RefChecks extends Transform {
               if (settings.warnNullaryOverride) {
-                if (other.paramss.isEmpty && !member.paramss.isEmpty && !member.isJavaDefined) {
+                def javaDetermined(sym: Symbol) = sym.isJavaDefined || isUniversalMember(sym)
+                if (other.paramss.isEmpty && !member.paramss.isEmpty && !javaDetermined(member) && !member.overrides.exists(javaDetermined))
```
This is what dot[ty|i] does. [I worry a little bit about the compiler efficiency of the `check allOverriddenSymbols.exists(test)`](https://contributors.scala-lang.org/t/nullary-and-nilary-methods/3756/10?u=lrytz) -- @retronym wdyt?

som-snytt replied:
I'll keep an eye on the grafana.

lrytz replied:
Looks OK

lrytz reviewed (commented):
LGTM! Let's try to merge this quickly to avoid another rebase (cc @retronym for the question ~above~below).

som-snytt commented on src/interactive/scala/tools/nsc/interactive/Global.scala:
```diff
@@ -193,23 +193,22 @@ class Global(settings: Settings, _reporter: Reporter, projectName: String = "")
-      super.addOne(binding)
-    }
-  }
+  */
```
Maybe address the todo.

som-snytt committed:
Cleaner unsafeWrapArray per review

som-snytt committed:
Avoid match warning

som-snytt committed:
Avoid shadowing vars with constructor params

som-snytt committed:
Annotate dangerous destructuring

som-snytt committed:
Unused imports, more unusedness

som-snytt committed:
Deprecations

Deprecation formality in use of deprecation.
Avoid procedure syntax.
Prefer hasPackageFlag.
Avoid any2stradd.
Avoid deprecated API.
Avoid procedure syntax.

som-snytt committed:
Avoid inferred Object and prefer path over toString

som-snytt committed:
Avoid inferred Object, prefer iterator

som-snytt committed:
Avoid missing-interpolator lint

lrytz commented:
🎉 thanks for pushing it through!
